hw3/zivzone/matrix.cpp

```diff
@@ -0,0 +1,193 @@
+  friend Matrix multiply_mkl(const Matrix &, const Matrix &);
+
+private:
+  size_t index(size_t row, size_t col) const
+    {
+        return row + col * m_nrow;
+    }
+
+  void reset_buffer(size_t nrow, size_t ncol)
+    {
+        if (m_buffer) { delete[] m_buffer; }
+        const size_t nelement = nrow * ncol;
+        if (nelement) { m_buffer = new double[nelement]; }
+        else          { m_buffer = nullptr; }
+        m_nrow = nrow;
+        m_ncol = ncol;
+    }
+  size_t m_nrow = 0;
+  size_t m_ncol = 0;
+  double * m_buffer = nullptr;
+
+  std::vector<double> matrix_data;
+};
+
+/*
+ * Naive matrix matrix multiplication.
+ */
+ /* 
+Matrix operator*(Matrix const & mat1, Matrix const & mat2)
+{
+    if (mat1.ncol() != mat2.nrow())
+    {
+        throw std::out_of_range(
+            "the number of first matrix column "
+            "differs from that of second matrix row");
+    }
+
+    Matrix ret(mat1.nrow(), mat2.ncol());
+
+    for (size_t i=0; i<ret.nrow(); ++i)
+    {
+        for (size_t k=0; k<ret.ncol(); ++k)
+        {
+            double v = 0;
+            for (size_t j=0; j<mat1.ncol(); ++j)
+            {
+                v += mat1(i,j) * mat2(j,k);
+            }
+            ret(i,k) = v;
+        }
+    }
+
+    return ret;
+}
+
+std::ostream & operator << (std::ostream & ostr, Matrix const & mat)
+{
+    for (size_t i=0; i<mat.nrow(); ++i)
+    {
+        ostr << std::endl << " ";
+        for (size_t j=0; j<mat.ncol(); ++j)
+        {
+            ostr << " " << std::setw(2) << mat(i, j);
+        }
+    }
+
+    return ostr;
+}
+*/
+
+Matrix multiply_naive(const Matrix &first_matrix, const Matrix &second_matrix) {
+  const size_t f_row = first_matrix.m_nrow;
+  const size_t f_col = first_matrix.m_ncol;
+  const size_t s_col = second_matrix.m_ncol;
+  Matrix result_matrix(f_row, s_col);
+  std::fill(std::begin(result_matrix.matrix_data), std::end(result_matrix.matrix_data), 0);
+  for (size_t i = 0; i < f_row; ++i){
+    for (size_t j = 0; j < s_col; ++j){
+      for (size_t k = 0; k < f_col; ++k){
+          result_matrix(i, j) += first_matrix(i, k) * second_matrix(k, j);
+      }
+    }
+  }
+    
+  return result_matrix;
+}
+
+Matrix multiply_mkl(const Matrix &first_matrix, const Matrix &second_matrix) {
+  size_t f_nrow = first_matrix.nrow();
+  size_t s_ncol = second_matrix.ncol();
+
+  //Matrix result_matrix(first_matrix.nrow(), second_matrix.ncol());
+  Matrix result_matrix(f_nrow, s_ncol);
+
+  std::fill(std::begin(result_matrix.matrix_data), std::end(result_matrix.matrix_data), 0);
+  cblas_dgemm(CblasRowMajor, CblasNoTrans, CblasNoTrans, 
+              first_matrix.nrow(), second_matrix.ncol(),first_matrix.ncol(), 1., first_matrix.matrix_data.data(), 
+              first_matrix.ncol(), second_matrix.matrix_data.data(), second_matrix.ncol(), 1., 
+              result_matrix.matrix_data.data(), result_matrix.ncol());
+  return result_matrix;
+}
+
+
+/* 
+int main(int argc, char ** argv)
```

Comment:
remember to remove comments if you want the reviewer can read your code clearly.